ethexe/processor/src/handling/run.rs

```diff
@@ -16,178 +16,282 @@
-        }
+    for (
+        actor_id,
+        StateHashWithQueueSize {
+            hash,
+            cached_queue_size,
+        },
+    ) in states
+    {
+        let queue_size = cached_queue_size as usize;
+        let chunk_idx = chunk_idx(queue_size, number_of_chunks);
+        chunks[chunk_idx].push((actor_id, hash, queue_size));
     }
+
+    chunks
+        .into_iter()
+        // Merge uneven chunks
+        .flatten()
+        // Repartition chunks in reverse order to ensure all chunks have an equal number of elements
+        .rev()
+        .chunks(chunk_size)
+        // Convert into vector of vectors
+        .into_iter()
+        .map(|c| c.into_iter().collect())
+        .collect()
 }
 
-async fn worker(
-    id: usize,
+fn run_runtime(
     db: Database,
-    instance_creator: InstanceCreator,
-    mut task_receiver: mpsc::Receiver<Task>,
-) {
-    log::trace!("Worker {} started", id);
+    executor: &mut InstanceWrapper,
+    program_id: ActorId,
+    state_hash: H256,
+) -> (ProgramJournals, H256) {
+    let code_id = db.program_code_id(program_id).expect("Code ID must be set");
 
-    let mut executor = instance_creator
-        .instantiate()
-        .expect("Failed to instantiate executor");
+    let instrumented_code = db.instrumented_code(ethexe_runtime::VERSION, code_id);
 
-    while let Some(task) = task_receiver.recv().await {
-        run_task(db.clone(), &mut executor, task).await;
-    }
+    executor
+        .run(db, program_id, code_id, state_hash, instrumented_code)
+        .expect("Some error occurs while running program in instance")
 }
 
-async fn one_batch(
-    from_index: usize,
-    task_senders: &[mpsc::Sender<Task>],
-    in_block_transitions: &mut InBlockTransitions,
-) -> BTreeMap<ProgramId, oneshot::Receiver<(Vec<JournalNote>, Option<Origin>)>> {
-    let mut result_receivers = BTreeMap::new();
+#[cfg(test)]
+mod tests {
+    use gprimitives::ActorId;
 
-    for (sender, (program_id, state_hash)) in task_senders
-        .iter()
-        .zip(in_block_transitions.states_iter().skip(from_index))
-    {
-        let (result_sender, result_receiver) = oneshot::channel();
+    use super::*;
 
-        let task = Task::Run {
-            program_id: *program_id,
-            state_hash: *state_hash,
-            result_sender,
-        };
+    #[test]
+    fn chunk_partitioning() {
+        const STATE_SIZE: usize = 1_000;
+        const CHUNK_PROCESSING_THREADS: usize = 16;
+        const MAX_QUEUE_SIZE: u8 = 20;
 
-        sender.send(task).await.unwrap();
+        let states = Vec::from_iter(
+            std::iter::repeat_with(|| {
+                (
+                    ActorId::from(0),
+                    StateHashWithQueueSize {
+                        hash: H256::zero(),
+                        cached_queue_size: (rand::random::<u8>() % MAX_QUEUE_SIZE + 1),
+                    },
+                )
+            })
+            .take(STATE_SIZE),
+        );
```

Comment:
```suggestion
        let states = vec![
              (
                  ActorId::from(0),
                  StateHashWithQueueSize {
                      hash: H256::zero(),
                      cached_queue_size: rand::random_range(1..=MAX_QUEUE_SIZE),
                  },
              ); STATE_SIZE
        ];
```

Reply:
It seems rand::random_range is missing in the current version and feature-gated in the modern one. I'll leave it as is.